src/v/base/format_to.h

```diff
@@ -67,10 +68,17 @@ struct formatter<T> {
 
 } // namespace fmt
 
+namespace std {
```

Comment:
Not saying we shouldn't do this, but AFAICT this is UB as putting things inside the `std` namespace is not allowed except for very limited carve-outs. It's the kind of UB that's "probably" fine though :shrug: .

Reply:
if the goal is to rid the tree of `operator<<` completely, then maybe its a fine risk if we commit to getting the job done before next release.

Reply:
http://google.github.io/googletest/advanced.html#teaching-googletest-how-to-print-your-values

Reply:
One way to fix it (since it's still going to be a problem due to boost/gtest and other things that want <<), is to bring the operator<< into the relevant namespace.

The problem occurs because the unqualified lookup never makes it to the global namespace, because it finds `operator<<` already in `ns` (even if it doesn't match), and then ADL looks in `std` and `ns` but not enclosing namespace (because ADL does not look in enclosing namespaces).

So we can bring the relevant `operator<<` into scope like so:

https://godbolt.org/z/evcPez7rE

The example compiles. In practice, you'd might want to have the concept operator<< in a details namespace so you can target it specifically.

Reply:
Yeah I actually like `using fmt::operator<<` (or the global namespace version).

Hrm let's just go with that then?

Reply:
gosh, tough choice. iiuc the decision is between the low-impact-technically-UB-but-probably-fine version that injects stuff into `std`, and scattering `using ::operator<<` all over the place? the UB version seems unique in that it's not injecting a few types, it's injecting maybe 1000 types into `std`, so maybe the compliant but verbose version is better.

do i understand correctly that if we pushed a magic button that generated fmt::formatters for every type we'd still have to add operator<<'s for googletest/boost, or could we solve the the test side separately?

Reply:
boost has less options, but I'm less concerned with it. We can define operator<< in `::std`, or `boost_test_print_type` in `::boost::test_tools::tt_detail`, or `boost_test_print_type` in `std`.

We might also be able to just add a template specialization for `print_log_value` in `::boost:test_tools::tt_detail` as well: https://github.com/boostorg/test/blob/097e97820e654ead9c477b47443a545cef5d3b12/include/boost/test/tools/detail/print_helper.hpp#L240

🤷

Reply:
having the `format_to` automatically creating the fmt::formatter is already awesome. i guess i'm wondering if having the operator<< automatically created is a big win? if it doesn't compile i'd just add one...?

Reply:
> i guess i'm wondering if having the operator<< automatically created is a big win? if it doesn't compile i'd just add one...?

Without `operator<<` Google test will just print the raw bytes of an object (mostly useless). Boosttet will fail to compile however. I do think for tests we want to enable printing somehow.

Reply:
i'm not saying we shoudln't have the operator<< i'm saying maybe we should just add it manually when needed like we do today?